src/openqaoa-core/utilities.py

```diff
@@ -1159,6 +1141,90 @@ def exp_val_hamiltonian_termwise(
+    See arXiv:2012.09738 and arXiv:2106.05800.
+
+    Parameters
+    ----------
+    hamiltonian: `Hamiltonian`
+        Hamiltonian object containing the problem statement.
+    calibration_measurements: `dict`
+        Dictionary containing the measurement counts of an empty QAOA circuit.
+    calibration_registers: `list`
+        List specifying the physical (device) qubits on which the calibration data has been obtained.
+        This is required because the calibration is usually performed over the whole device and hence the measurement outcomes (the calibration data) are strings with the size of the whole device while usually only a particular section is used.
+    qubit_mapping: `list`
+        List specifying the physical (device) qubits on which the QAOA circuit is executed. Related to qubit selection and qubit routing.
+
+    Returns
+    -------
+    calibration_factors: `dict`
+        Calibration factors as a dict.
+    """
+    calibration_registers_dict = {v: k for k, v in enumerate(calibration_registers)}
+
+    # Define number of qubits, problem hamiltonian and QAOA parameters
+    n_qubits = hamiltonian.n_qubits
+
+    # Extract Hamiltonian terms
+    terms = list(hamiltonian.terms)
+
+    if qubit_mapping == None:
+        qubit_mapping = np.arange(0, n_qubits)
+
+    # Initialize an empty dict
+    calibration_factors = {}
+
+    # Compute single spin and pairs of spins expectation values of terms present in the Hamiltonian
+    for term in terms:
+        # If bias term compute single spins expectation value
+        if len(term) == 1:
+            i = term.qubit_indices[0]
+            i_phys = qubit_mapping[i]
+            i_cal = calibration_registers_dict[i_phys]
+            exp_val_z = exp_val_single(i_cal, calibration_measurements)
+            calibration_factors.update({(i,): exp_val_z})
+
+        # If two-body term compute pairs of spins expectation values
+        elif len(term) == 2:
+            i, j = term.qubit_indices  # problem indices, ex: (0,1)
+            i_phys, j_phys = (
+                qubit_mapping[i],
+                qubit_mapping[j],
+            )  # physical indices, ex: (133, 131) after routing
+            i_cal, j_cal = (
+                calibration_registers_dict[i_phys],
+                calibration_registers_dict[j_phys],
+            )  # calibration indices, i.e. to which location on the measurement string each physical qubit corresponds to, ex: (63, 61)
+            exp_val_zz = exp_val_pair((i_cal, j_cal), calibration_measurements)
+
+            calibration_factors.update(
+                {(i, j): exp_val_zz}
+            )  # calibration factors are calculated for the terms in the hamiltonian/problem
+
+        # If constant term, ignore
+        if len(term) == 0:
```

Comment:
Should this if be on its own? Or is it part of the previous conditional (and it should be an else or something)?

Reply:
changed